extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/SplitPackageProcessor.java

```diff
@@ -0,0 +1,53 @@
+package io.quarkus.arc.deployment;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.jboss.jandex.ClassInfo;
+import org.jboss.logging.Logger;
+
+import io.quarkus.arc.processor.DotNames;
+import io.quarkus.deployment.ApplicationArchive;
+import io.quarkus.deployment.annotations.BuildProducer;
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.ApplicationArchivesBuildItem;
+
+/**
+ * Split package (same package coming from multiple app archives) is considered a bad practice and
+ * this processor tries to detect it and log a warning listing out the offending packages.
+ * <p>
+ * Note that this processor is a best-effort because it only operates on {@code ApplicationArchivesBuildItem} which
+ * means that if a 3rd party library isn't indexed, we aren't able to detect it even though it can still be a part of
+ * resulting application. See also {@code io.quarkus.arc.processor.BeanArchives.IndexWrapper}.
+ */
+public class SplitPackageProcessor {
+
+    private static final Logger LOGGER = Logger.getLogger(SplitPackageProcessor.class);
+
+    @BuildStep
+    void splitPackageDetection(ApplicationArchivesBuildItem archivesBuildItem,
+            // Dummy producer to make sure the build step executes
+            BuildProducer<ValidationPhaseBuildItem.ValidationErrorBuildItem> dummy) {
+        Set<String> allKnownPackages = new HashSet<>();
+        Set<String> splitPackages = new HashSet<>();
+        // for all app archives
+        for (ApplicationArchive archive : archivesBuildItem.getAllApplicationArchives()) {
+            Set<String> singleArchivePackages = new HashSet<>();
+            // and for each known class in each archive
+            for (ClassInfo classInfo : archive.getIndex().getKnownClasses()) {
+                singleArchivePackages.add(DotNames.packageName(classInfo.name()));
+            }
+            // perform intersection
+            Set<String> singleArchiveSplitPackages = new HashSet<>(singleArchivePackages);
+            singleArchiveSplitPackages.retainAll(allKnownPackages);
+            // add single archive packages to set of all known packages
+            allKnownPackages.addAll(singleArchivePackages);
+            // and remember all split packages
+            splitPackages.addAll(singleArchiveSplitPackages);
+        }
+        if (!splitPackages.isEmpty()) {
+            LOGGER.warnf("Detected a split package usage which is considered a bad practice and should be avoided. " +
```

Comment:
@mkouba Not sure I understand. Do you mean listing out all classes which have split package issues?
I thought that might get too verbose because you can easily have 50 classes inside given package in your `app` and only a singular class in `lib` (with same package) but you end up listing 51 classes.

Reply:
I mean something like:
```
Detected a split package usage which is considered a bad practice and should be avoided. Following packages were found in multiple archives: 
- "com.me.app.sub" found in [archiveA, archiveB]
- "org.acme.foo" found in [archiveC, archiveB]
```